src/libraries/System.Private.CoreLib/src/System/DateTime.Windows.cs

```diff
@@ -15,31 +15,26 @@ public static unsafe DateTime UtcNow
         {
             get
             {
-                ulong fileTime;
-                s_pfnGetSystemTimeAsFileTime(&fileTime);
+                ulong fileTimeTmp; // mark only the temp local as address-taken
+                s_pfnGetSystemTimeAsFileTime(&fileTimeTmp);
+                ulong fileTime = fileTimeTmp;
 
                 if (s_systemSupportsLeapSeconds)
                 {
-                    Interop.Kernel32.SYSTEMTIME time;
-                    ulong hundredNanoSecond;
+                    // Query the leap second cache first, which avoids expensive calls to GetFileTimeAsSystemTime.
 
-                    if (Interop.Kernel32.FileTimeToSystemTime(&fileTime, &time) != Interop.BOOL.FALSE)
+                    LeapSecondCache cacheValue = s_leapSecondCache;
+                    ulong ticksSinceStartOfCacheValidityWindow = fileTime - cacheValue.OSFileTimeTicksAtStartOfValidityWindow;
+                    if (ticksSinceStartOfCacheValidityWindow < LeapSecondCache.ValidityPeriodInTicks)
                     {
-                        // to keep the time precision
-                        ulong tmp = fileTime; // temp. variable avoids double read from memory
-                        hundredNanoSecond = tmp % TicksPerMillisecond;
-                    }
-                    else
-                    {
-                        Interop.Kernel32.GetSystemTime(&time);
-                        hundredNanoSecond = 0;
+                        return new DateTime(dateData: cacheValue.DotnetDateDataAtStartOfValidityWindow + ticksSinceStartOfCacheValidityWindow);
                     }
 
-                    return CreateDateTimeFromSystemTime(in time, hundredNanoSecond);
+                    return UpdateLeapSecondCacheAndReturnUtcNow(); // couldn't use the cache, go down the slow path
```

Comment:
>UpdateLeapSecondCacheAndReturnUtcNow [](start = 27, length = 36)

nit: should we make this method take the parameter which we can pass "fileTimeTmp" value and avoid having the method to call s_pfnGetSystemTimeAsFileTime again?

Reply:
Tried this, but the codegen at the caller side ended up being worse. I think it's an acceptable tradeoff to make the "cache hit" case as fast as possible even if it makes the "cache miss" case a little slower. Do you have thoughts on this?

Reply:
If the codegen is worse then, I am fine with what you have as this will be called in the slower path.